### ISSUE TITLE
Exercise Outcome 

### ISSUE THREAD

**Issue by tfnribeiro** [zeeguu-ecosystem/Zeeguu-API#304]
The other day we were discussing exercise outcomes and I realized that there is a bit of a disconnect between how we consider an outcome.

At the moment, an outcome is associated with an exercise instance, which collects some of the user interactions during the exercise, such as Hints, Translations, Wrong attempts. Each exercise is then associated with a bookmark, the one being tested in the bookmark_exercise_mapping table.

However, this creates an issue where if we consider exercises that have multiple bookmarks, e.g. Match, MultipleChoiceContext. In these cases, I believe we should still keep the bookmarks we presented in that exercise with a mapping and the outcome is tied to each bookmark rather than the exercise as a whole. For example, in Match we currently can only tell which bookmarks were shown together by looking at the time in the exercises and it means we have 2 exercise ids for the same exercise instance. 

I would propose associating the outcome with the bookmark_exercise_mapping rather than the exercise itself, as the outcomes have a direct impact on the bookmark progression. Bookmarks that aren't tested can be given a NULL outcome, so we know they were distractors.

**Comment by mircealungu** [comment]
Very good observation! 

I wonder if this is an opportunity to rename that `bookmark_exercise_mapping` table to something more semantically meaningful. E.g. `bookmark_exercise_encounter`? 

I was teaching in TID last Friday, and I had this epiphany (more like a recollection of something that I've heard before) that when you have a "mapping" table it's a missed opportunity for thinking deeper what is this concept that is being captured actually!

**Comment by mircealungu** [comment]
In this particular case, if we model the `encounter` then it would become natural to add a column that would track the role of the bookmark in the encounter: distractor, main word being tested, etc.

**Comment by tfnribeiro** [comment]
I agree with that, ultimately this would allow us to re-create exercises with the same bookmarks shown, which could also be good for other cases - for example sharing exercises if we ever want to do it.

**Comment by tfnribeiro** [comment]
The only issue I see is that we can't really do backwards compatability, besides including the words we already have mapped to an exercise. E.g. We won't know what words were included in Context Exercises, or the 3rd word in the Match exercise as those are currently not logged in the system. 

Ideally, we would be able to migrate the changes to this "new" way of modeling the encounters/outcomes, or we risk introducing ever more confusion on how to interpret the data in the exercises.
